File: Python/algoritms/leetCode/169-majority_element.py
import unittest


def majorityElement(nums: list[int]) -> int:
    nums_dic = {}
    num_max = {"max": 0, "key": 0}
    for i in nums:
        nums_dic.setdefault(i, 0)
        if i in nums_dic:
            nums_dic[i] += 1

    for k, v in nums_dic.items():
        if v > num_max["max"]:
            num_max["max"] = v
            num_max["key"] = k
    return num_max["key"]


def majorityElement2(nums: list[int]) -> int:
    res, count = 0, 0

    for n in nums:
        if count == 0:
            res = n
        count += (1 if n == res else -1)
    return res


def majorityElement3(nums: list[int]) -> int:
    count = {}
    res, maxCount = 0, 0

    for n in nums:
        count[n] = 1 + count.get(n, 0)
        res = n if count[n] > maxCount else res
        maxCount = max(count[n], maxCount)
    return res


class TestMajorityElement(unittest.TestCase):

    def TestMajorityElement1(self):
        n = majorityElement([3, 3, 4, 4, 3])
        self.assertEqual(n, 3)

    def TestMajorityElement2(self):
        n = majorityElement2([2, 2, 3])
        self.assertEqual(n, 2)

    def TestMajorityElement3(self):
        n = majorityElement3([1, 1, 4])
        self.assertEqual(n, 1)


if __name__ == "__main__":
    unittest.main()

File: TypeScript/leetcode/169-majorityElement.ts
import { assertEquals } from "https://deno.land/std@0.192.0/testing/asserts.ts";

function majorityelement1(nums: number[]): number {
  const numMap: Map<number, number> = new Map();
  const numMax: Map<string, number> = new Map();
  numMax.set("key", 0);
  numMax.set("max", 0);

  for (const n of nums) {
    numMap.set(n, (numMap.get(n)! || 0) + 1);
  }
  for (const [key, value] of numMap) {
    if (value > numMax.get("max")!) {
      numMax.set("max", value);

      numMax.set("key", key);
    }
  }
  return numMax.get("max")!;
}

function majorityelement2(nums: number[]): number {
  let res = 0;
  let count = 0;

  for (const n of nums) {
    if (count === 0) {
      res = n;
    }
    count += n === res ? 1 : -1;
  }
  return res;
}

function majorityelement3(nums: number[]): number {
  const count: Map<number, number> = new Map();
  let res = 0;
  let maxCount = 0;
  for (const n of nums) {
    count.set(n, (count.get(n) || 0) + 1);
    res = count.get(n)! > maxCount ? n : res;
    maxCount = Math.max(count.get(n)!, maxCount);
  }
  return res;
}

Deno.test("majorityelement1", () => {
  const nums = [1, 2, 2];
  const n = majorityelement1(nums);
  const n2 = majorityelement2(nums);
  const n3 = majorityelement3(nums);
  assertEquals(n, 2);
  assertEquals(n3, 2);
  assertEquals(n2, 2);
});
